### PR TITLE
Prometheus: Feedback link for metric encyclopedia plus performance improvements

### LINKED ISSUE #64581: show # of results by results header.

### LINKED ISSUE #64582: keep total metrics for backend search in top, show backend filter by Results header.

### LINKED ISSUE #64583: add feedback option for users to complete a form and opt in for an interview.

### LINKED ISSUE #64621: manually backport MetricEncyclopediaModal.tsx to 9.4.x
https://github.com/grafana/grafana/pull/64575#event-8717477883

Correctly debounce the uFuzzy fuzzy search

### PR DESCRIPTION
**Pull request by bohandley**
Fixes https://github.com/grafana/grafana/issues/64581
Fixes https://github.com/grafana/grafana/issues/64582
Fixes https://github.com/grafana/grafana/issues/64583
Fixes https://github.com/grafana/grafana/issues/64621

Screenshot 2023-03-10 at 2 36 42 PM

This adds a feedback link for the metric encyclopedia. We will be asking UX, SEs, SREs and community members to review it under the feature flag. 

It also improves performance by correctly debouncing the uFuzzy search (thank you @leeoniya) and reduces calls to filterMetrics().

### CODE REVIEW

**Review by gtk-grafana — CHANGES_REQUESTED** [review]
Some weirdness with the filters I mentioned on slack, everything else is working good.

File: public/app/plugins/datasource/prometheus/querybuilder/components/MetricEncyclopediaModal.tsx [hunk 1]
```diff
@@ -260,14 +263,14 @@ export const MetricEncyclopediaModal = (props: Props) => {
         // backend and frontend
         // skipLetterSearch is purely for setting letter spans as active so they can be clicked
         // where we don't filter out by letter search and keep letters of all other filters
-        if (letterSearch && !skipLetterSearch) {
+        if ((keepMetric || !fuzzySearchQuery) && letterSearch && !skipLetterSearch) {
```
Comment (bohandley) [inline public/app/plugins/datasource/prometheus/querybuilder/components/MetricEncyclopediaModal.tsx:266]: @gtk-grafana I need to come up with a set of rules for the filtering. That will make it easier to keep track of the order of things. For now, it's 1) fuzzy search, 2) alphabet, 3) filter types. I also need to consider the metrics that don't have metadata, they are not filtered out by type unless the switch is on to filter them.

**Comment by bohandley** [comment]
> Some weirdness with the filters I mentioned on slack, everything else is working good.

Made the change!

**Comment by leeoniya** [comment]
i'm getting 472 results for "prom" and 1067 results for "prome"? :thinking: 

image
image

**Comment by bohandley** [comment]
> i'm getting 472 results for "prom" and 1067 results for "prome"? 🤔
> 
> image image

@leeoniya I believe this is because uFuzzy sees prome as a mispelling of process and so includes those results as well. My understanding of uFuzzy is that it returns results for close spellings

"prom" -> prometheus

but         
                 -> prometheus
"prome" 
                 -> process

**Comment by bohandley** [comment]
Filters are still broken. I need to fix those.

**Comment by bohandley** [comment]
> ah, that makes sense, we just need to do better with sorting by relevance.
> 
> `uf.filter(haystack, query);` does no sorting, we gotta switch to the `uf.search()` API.

So we can sort results based on the uFuzzy search api. Currently results are sorted alphabetically. I can add an item to the epic to add sorting rules
1) alphabetically
2) uFuzzy sort relevance

**Review by gtk-grafana — APPROVED** [review]
Now the metrics free text field and the type filters interact as I expect, but I'm still seeing some confusing behavior with the filters.

I assume that within the type filter, that each category is additive, but when I select all categories, I do not get the same value as no categories. Maybe there are some values without any category? I think it's that there are a bunch of values without metadata and I see that you added some UI for that. 

I propose we remove the toggle for excluding metadata, and instead add another category of "no metadata" or "uncategorized" or something within the filter element, and make sure that they don't show up when the other filters are displayed, otherwise I see all the result numbers that aren't adding up and perceive that the UI is broken, or hiding something from me.

That said, this issue already exists in the main branch, so I won't let this block merge, as this brings much needed improvements to the feature!

**Comment by bohandley** [comment]
> Now the metrics free text field and the type filters interact as I expect, but I'm still seeing some confusing behavior with the filters.
> 
> I assume that within the type filter, that each category is additive, but when I select all categories, I do not get the same value as no categories. Maybe there are some values without any category? I think it's that there are a bunch of values without metadata and I see that you added some UI for that.
> 
> I propose we remove the toggle for excluding metadata, and instead add another category of "no metadata" or "uncategorized" or something within the filter element, and make sure that they don't show up when the other filters are displayed, otherwise I see all the result numbers that aren't adding up and perceive that the UI is broken, or hiding something from me.
> 
> That said, this issue already exists in the main branch, so I won't let this block merge, as this brings much needed improvements to the feature!

There are a bunch of metrics with no type or description at all! I will look further into this behavior though. I am always including metrics with no type because I believe an admin could have created a metric without a type, an SRE could know they need a counter, they look it up with the counter type selected, and because it was never given metadata it won't show up. I can see this as being weird though.

I'll look into the results adding up too!

Thank you @gtk-grafana, your reviews are excellent! 🎉

**Pull request merged**